core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,773 @@
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = dcp_lock_channel(&data->dcp_data.channel);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	desc = &data->dcp_data.desc;
+
+	desc->ctrl0 = DCP_CONTROL0_DECR_SEMAPHORE | DCP_CONTROL0_ENABLE_CIPHER |
+		      DCP_CONTROL0_INTERRUPT_ENABLE;
+	desc->ctrl1 = DCP_CONTROL1_CIPHER_SELECT_AES128;
+
+	if (init->op == DCP_ENCRYPT)
+		desc->ctrl0 |= DCP_CONTROL0_CIPHER_ENCRYPT;
+
+	if (init->key_mode == DCP_OTP) {
+		desc->ctrl0 |= DCP_CONTROL0_OTP_KEY;
+		desc->ctrl1 |= DCP_CONTROL1_KEY_SELECT_OTP_CRYPTO;
+	} else if (init->key_mode == DCP_PAYLOAD) {
+		desc->ctrl0 |= DCP_CONTROL0_PAYLOAD_KEY;
+		if (!init->key)
+			return TEE_ERROR_BAD_PARAMETERS;
+		memcpy(data->key, init->key, DCP_AES128_KEY_SIZE);
+	} else {
+		desc->ctrl1 |= SHIFT_U32(init->key_mode, 8);
+	}
+
+	if (init->mode == DCP_CBC) {
+		desc->ctrl0 |= DCP_CONTROL0_CIPHER_INIT;
+		desc->ctrl1 |= DCP_CONTROL1_CIPHER_MODE_CBC;
+		if (!init->iv)
+			return TEE_ERROR_BAD_PARAMETERS;
+		memcpy(data->iv, init->iv, DCP_AES128_IV_SIZE);
+	}
+
+	/* Allocate aligned buffer for dcp iv and key */
+	ret = dcp_calloc_align_buf(&data->payload,
+				   DCP_AES128_IV_SIZE + DCP_AES128_KEY_SIZE);
+	if (ret != TEE_SUCCESS)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	desc->src_buffer = NULL;
+	desc->dest_buffer = NULL;
+	desc->status = 0;
+	desc->buff_size = 0;
+	desc->next = virt_to_phys(desc);
+
+	data->initialized = true;
+
+	return ret;
+}
+
+TEE_Result dcp_cipher_do_update(struct dcp_cipher_data *data,
+				const uint8_t *src, uint8_t *dst, size_t size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct dcp_align_buf output = { };
+	struct dcp_align_buf input = { };
+	struct dcp_descriptor *desc = NULL;
+
+	if (!data || !src || !dst)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!data->initialized) {
+		EMSG("Error, please call dcp_aes_do_init() before");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (size % DCP_AES128_BLOCK_SIZE) {
+		EMSG("Input size has to be a multiple of %zu bytes",
+		     DCP_AES128_BLOCK_SIZE);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = dcp_calloc_align_buf(&output, size);
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	ret = dcp_calloc_align_buf(&input, size);
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	desc = &data->dcp_data.desc;
+
+	/* Copy input data */
+	memcpy(input.data, src, size);
+
+	/* Copy key and IV */
+	memcpy(data->payload.data, data->key, DCP_AES128_KEY_SIZE);
+	data->payload_size = DCP_AES128_KEY_SIZE;
+	if (desc->ctrl0 & DCP_CONTROL0_CIPHER_INIT) {
+		memcpy(data->payload.data + DCP_AES128_KEY_SIZE, data->iv,
+		       DCP_AES128_IV_SIZE);
+		data->payload_size += DCP_AES128_IV_SIZE;
+	}
+
+	desc->src_buffer = (uint8_t *)input.paddr;
+	desc->dest_buffer = (uint8_t *)output.paddr;
+	desc->payload = (uint8_t *)data->payload.paddr;
+	desc->buff_size = size;
+
+	cache_operation(TEE_CACHECLEAN, data->payload.data, data->payload_size);
+	cache_operation(TEE_CACHECLEAN, input.data, size);
+	cache_operation(TEE_CACHECLEAN, output.data, size);
```

Comment:
To me it seems that using `TEE_CACHEINVALIDATE` here instead and skipping the `TEE_CACHEINVALIDATE` below should be enough.
No matter what we'll do there will be race with the cache if someone accesses the `output.data` buffer while `dcp_run()` is executed.

Reply:
I agree, a ```TEE_CACHEINVALIDATE``` is enough